+ show the inferred exception list (as part of the type) for functions that don't have an explicit `.raises` pragma

nickysn opened the pull request.

nickysn committed:
+ show the inferred exception list (as part of the type) for functions that don't have an explicit `.raises` pragma

nickysn committed:
+ update tests to include the new `.raises: <inferred>` proc type info

nickysn committed:
* fixed test (was missed in my previous commit)

nickysn committed:
+ also update nimsuggest tests with `.raises: <inferred>`

nickysn committed:
* fixed some tests that were missed in the previous commit

arnetheduck commented:
maybe introduce a syntax for inferred / augmented information, ie `{.*raises: [..]*, ....}` or something like that? 

that said, generally, inferred raises has less information value as it doesn't represent a deliberate choice by the author of the API, so it should also probably go behind everything else that was explicitly included in the signature (if a function has more pragmas)

nickysn commented:
`.*raises: [..]*` looks way more cryptic than `.raises: <inferred> [..]`. Will people know what it means?

Moving it last, after the explicitly set pragmas is fine with me. I'll update the pull request.

arnetheduck commented:
ideally it would appear in italic or some other font, which * is in markdown - not sure what the possibilities are ... ie type inlay hints appear greyed in rust in vscode for example.

nickysn committed:
* the `raises: <inferred>` exception list moved last, after all the pragmas that were explicitly defined by the programmer

nickysn commented:
`.*raises: [..]*` looks way more cryptic than `.raises: <inferred> [..]`. Will people know what it means?

Moving it last, after the explicitly set pragmas is fine with me. I'll update the pull request.

> ideally it would appear in italic or some other font, which * is in markdown - not sure what the possibilities are ... ie type inlay hints appear greyed in rust in vscode for example.

It doesn't work like that. First, for inlay hints, we have no control over the color of the text, it is decided by the VS code color theme. The Nim inlay type hints and the Rust inlay type hints are exactly the same color, for this particular reason.

But this isn't shown in an inlay hint, this is shown on hover, and it is indeed colored, but it follows the colors for Nim syntax highlighting, and is not interpreted as markdown.

Here's how it would look with `.*raises: [..]*`:
![image](https://github.com/nim-lang/Nim/assets/22420132/c577f91b-2158-455d-bb09-4bac10fc4995)

And here's how it looks with `.raises: <inferred> [..]`:
![image](https://github.com/nim-lang/Nim/assets/22420132/96d2cd2e-2ad4-439d-80f9-8ee6783ca656)

This info will also be shown on the console as part of compiler error messages and the console also doesn't support markdown.

nickysn committed:
* fixed tests that were missed in my previous commit

nickysn committed:
* fixed yet another test

Araq commented on compiler/types.nim:
```diff
@@ -723,20 +723,35 @@ proc typeToString(typ: PType, prefer: TPreferedDesc = preferName): string =
       if tfThread in t.flags:
         addSep(prag)
         prag.add("gcsafe")
+      if not hasImplicitRaises and not isNil(t.owner) and not isNil(t.owner.typ) and not isNil(t.owner.typ.n) and (t.owner.typ.n.len > 0):
```
Needs to be opt-in via the `prefer` parameter as in general the error messages became more spammy and only nimsuggest wants it.

nickysn replied:
Ok, done. Introduced preferInferredEffects. The error messages are unchanged now, only nimsuggest hints use it.

nickysn committed:
+ introduced TPreferedDesc.preferInferredEffects. Only show the
  `.raises: <inferred>` pragma, when this prefer type is used. Only use this in
  nimsuggest for now, not for error messages.

nickysn committed:
Merge branch 'devel' into inferred_exceptions_hint
